Fixed copy issues and improved functionalities.

Ergus opened the pull request:
There were some issues in the previous patch that were fixed with the highlight when copy. 

Added another configuration option for the copy function to avoid highlight after preselected region (keep the original copy-as-kill behaviour).

Modified the which-key integration to use an idle-delay to be more consistent with the which-key behavior.

Ergus committed:
Bugfix in overlay.

Ergus committed:
Fix consecutive copies issue.

Ergus committed:
Added which-key delay.

This uses the idle delay of which-key in the same way which-key
does. This is more coherent with which-key default behavior.

Ergus committed:
Improved the kill function.

- Now the highlight is not created every time the command repeats,
just moved.

(A deleted overlay is not permanently disconnected. You can give it a
position in a buffer again by calling move-overlay.)

- There is a new custom variable composable-kill-region-highlight to
enable/disable the post highlight when calling a composable command with an
active region.

Ergus committed:
Added documentation for the highlight region after copy.

Ergus committed:
Deleted unused face.

Ergus committed:
Fixed byte compiler warning.

Ergus committed:
Some changes in the pairs.

Simplification to avoid unneeded funcal and extra function just used once.

Ergus committed:
Added # to symbols names.

Ergus committed:
Added composable-goto-char function.

This brings a functionality like the f action in vim.
so now C-w c [char] copy the region until the next occurrence of
[char].
Prefix works as expected.
TODO: There is an issue when canceling with C-g during the char prompt.

s-kostyaev commented:
@Ergus in this version (and in latest upstream version) copy of already marked region is broken:
```
Debugger entered--Lisp error: (error "Marker points into wrong buffer" #<marker in no buffer>)
  make-overlay(#<marker in no buffer> 5249)
  (let ((o (make-overlay composable--start-point point))) (if (and (> composable--count 1) composable-repeat-copy-save-last) (progn (setq last-command 'kill-region))) (copy-region-as-kill mark point) (setq composable--overlay o) (overlay-put o 'priority 999) (overlay-put o 'face 'composable-highlight) (add-hook 'pre-command-hook 'composable--delete-highlight))
  (closure (t) (mark point) (interactive (list (mark) (point))) (let ((o (make-overlay composable--start-point point))) (if (and (> composable--count 1) composable-repeat-copy-save-last) (progn (setq last-command 'kill-region))) (copy-region-as-kill mark point) (setq composable--overlay o) (overlay-put o 'priority 999) (overlay-put o 'face 'composable-highlight) (add-hook 'pre-command-hook 'composable--delete-highlight)))(5200 5249)
  funcall-interactively((closure (t) (mark point) (interactive (list (mark) (point))) (let ((o (make-overlay composable--start-point point))) (if (and (> composable--count 1) composable-repeat-copy-save-last) (progn (setq last-command 'kill-region))) (copy-region-as-kill mark point) (setq composable--overlay o) (overlay-put o 'priority 999) (overlay-put o 'face 'composable-highlight) (add-hook 'pre-command-hook 'composable--delete-highlight))) 5200 5249)
  call-interactively((closure (t) (mark point) (interactive (list (mark) (point))) (let ((o (make-overlay composable--start-point point))) (if (and (> composable--count 1) composable-repeat-copy-save-last) (progn (setq last-command 'kill-region))) (copy-region-as-kill mark point) (setq composable--overlay o) (overlay-put o 'priority 999) (overlay-put o 'face 'composable-highlight) (add-hook 'pre-command-hook 'composable--delete-highlight))))
  (cond ((region-active-p) (call-interactively command)) (composable-object-mode (setq this-command composable-twice-mark) (funcall composable-twice-mark arg)) (t (setq composable--command-prefix arg composable--command command) (composable-object-mode)))
  composable-save-region(nil)
  funcall-interactively(composable-save-region nil)
  call-interactively(composable-save-region nil nil)
  command-execute(composable-save-region)
```

Ergus committed:
Big changes:

* Added a composable-mark-paragraph.
* Fixed C-SPC C-SPC in composable-mark-mode, it was removing the mark,
but not disabling the composable-object-mode.
* Fixed the repeat commands like composable-mark-*
* Added a new composable--border-point boundary variable to remember
the region bounds in mark commands for a right overlay.
* Moved some commands to composable--start.
* Fixed the half-cursor function call and variable initialization.

Ergus committed:
Added some tests for the new composable goto-char function.

Ergus commented:
Hi I just pushed some  many new changes/ fixes, but they are ready to merge. All the tests run (except the C-g one, but it works in real life, so I thinks there is an issue with the C-g emulation in cask/ecukes).
And also some downloads failed in travis.
Please check the changes in the local machine and add them if they work for you. Specially the first commits that fixes an important issue.

Ergus commented:
Hi: 
@paldepind have you had time to test this changes? Because the current version in melpa is broken. If any user tries it now he will get a very bad impression about it.

Ergus committed:
Fix and improvement.

Fix first call overlay issue that was failing.

Fix the mode-line color change that was failing in some race
conditions. Now it is simpler and faster.

Remove the extra hook composable--delete-highlight. This keeps the
same original emacs behavior, and reduces some operations and
simplifies the code to behave better.

Ergus commented:
Hi I made 3 other important fixes and optimizations. Please if you don't have time to check or maintain this could you transfer the ownership of the melpa repo to me and I will take care of it?

Ergus committed:
Workaround to avoid issues with multiple-cursors-mode.
